File: alg_ds_1/linked_list_2/linked_list_2.go
package main

import (
	"errors"
	"os"
	"reflect"
)

type Node struct {
	next  *Node
	prev  *Node
	value int
}

type LinkedList2 struct {
	head *Node
	tail *Node
}

func (l *LinkedList2) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
		l.head.next = nil
		l.head.prev = nil
	} else {
		l.tail.next = &item
		item.prev = l.tail
	}

	l.tail = &item
	l.tail.next = nil
}

func (l *LinkedList2) Count() int {
	count := 0
	cur := l.head
	for cur != nil {
		count++
		cur = cur.next
	}
	return count
}

// error не nil, если узел не найден
func (l *LinkedList2) Find(n int) (Node, error) {
	curr := l.head
	for curr != nil {
		if curr.value == n {
			return *curr, nil
		}
		curr = curr.next
	}
	return Node{}, errors.New("node not found")
}

func (l *LinkedList2) FindAll(n int) []Node {
	var nodes []Node
	curr := l.head
	for curr != nil {
		if curr.value == n {
			nodes = append(nodes, *curr)
		}
		curr = curr.next
	}
	return nodes
}

func (l *LinkedList2) Delete(n int, all bool) {
	if l.head == nil {
		return
	}
	current := l.head
	for current != nil {
		if current.value == n {
			if current.prev != nil {
				current.prev.next = current.next
			} else {
				l.head = current.next
			}
			if current.next != nil {
				current.next.prev = current.prev
			} else {
				l.tail = current.prev
			}
			if !all {
				return
			}
		}
		current = current.next
	}
}

func (l *LinkedList2) Insert(after *Node, add Node) {
	newNode := &add

	if after == nil {
		newNode.next = l.head
		newNode.prev = nil
		if l.head != nil {
			l.head.prev = newNode
		}
		l.head = newNode
		if l.tail == nil {
			l.tail = newNode
		}
		return
	}
	newNode.prev = after
	newNode.next = after.next

	after.next = newNode

	if newNode.next != nil {
		newNode.next.prev = newNode
	} else {
		l.tail = newNode
	}
}

func (l *LinkedList2) InsertFirst(first Node) {
	if l.head == nil {
		l.head = &first
		l.tail = &first
		return
	}
	first.next = l.head
	l.head.prev = &first
	l.head = &first
}

func (l *LinkedList2) Clean() {
	l.head = nil
	l.tail = nil
}

File: alg_ds_1/python_version/02_linked_list_2/linked_list_2.py
class Node:

    def __init__(self, v):
        self.value = v
        self.prev = None
        self.next = None


class LinkedList2:

    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
            item.prev = None
            item.next = None
        else:
            self.tail.next = item
            item.prev = self.tail
        self.tail = item

    # Задание на курсе: 2
    # Задача: 2.1
    # Название: поиск первого узла по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def find(self, val):
        current = self.head

        while current is not None:
            if current.value == val:
                return current
            current = current.next

        return None

    # Задание на курсе: 2
    # Задача: 2.2
    # Название: поиск всех узлов по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(n)
    def find_all(self, val):
        result = []
        current = self.head

        while current is not None:
            if current.value == val:
                result.append(current)
            current = current.next

        return result

    # Задание на курсе: 2
    # Задачи: 2.3, 2.4
    # Название: удаление первого или всех узлов по значению
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def delete(self, val, all=False):
        current = self.head
        while current is not None:
            next_node = current.next
            if current.value == val:
                if current.prev is None:
                    self.head = current.next
                else:
                    current.prev.next = current.next
                if current.next is None:
                    self.tail = current.prev
                else:
                    current.next.prev = current.prev
                if not all:
                    return
            current = next_node
    # Задание на курсе: 2
    # Задача: 2.7
    # Название: очистка списка
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def clean(self):
        self.head = None
        self.tail = None

    # Задание на курсе: 2
    # Задача: 2.8
    # Название: вычисление длины списка
    # Временная сложность: O(n)
    # Пространственная сложность: O(1)
    def len(self):
        count = 0
        current = self.head

        while current is not None:
            count += 1
            current = current.next

        return count

    # Задание на курсе: 2
    # Задача: 2.5
    # Название: вставка узла после заданного узла
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def insert(self, afterNode, newNode):
        if afterNode is None:
            self.add_in_head(newNode)
            return
        newNode.prev = afterNode
        newNode.next = afterNode.next
        if afterNode.next is None:
            self.tail = newNode
        else:
            afterNode.next.prev = newNode
        afterNode.next = newNode

    # Задание на курсе: 2
    # Задача: 2.6
    # Название: добавление узла в голову списка
    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def add_in_head(self, newNode):
        newNode.prev = None
        newNode.next = self.head
        if self.head is None:
            self.tail = newNode
        else:
            self.head.prev = newNode
        self.head = newNode
